operator/pkg/functional/compare.go

```diff
@@ -0,0 +1,46 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package functional
+
+func CompareMaps[T, U comparable](a, b map[T]U) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for key := range a {
+		if a[key] != b[key] {
+			return false
+		}
+	}
+	return true
+}
+
+func CompareMapsFn[T comparable, U any](a, b map[T]U, fn func(U, U) bool) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for key := range a {
+		if !fn(a[key], b[key]) {
+			return false
+		}
+	}
+	return true
+}
+
+func CompareConvertibleSlices[T, U any](a []T, b []U, fn func(T, U) bool) bool {
```

Comment:
Does `slices.EqualFunc` not work here? The implementations appear to be the same.

Similar to our short discussion about `require` and `assert`, it'd be nice to utilize the stdlib when possible for consistency purposes. If we need to diverge or extend the packages can we leave a record of what's different and/or why?

Reply:
It does, I just removed most of these in lieu of heavier use of built-in `slices` functions and `reflect.DeepEqual`.